### ISSUE TITLE
ValueError resulting from dict object being passed to ast.literal_eval() by _convert_string_to_native()

### ISSUE THREAD

**Issue by JessicaAlan** [foremast/foremast#326]
After upgrading from 3.46.1 to 3.51.7 users started getting this error

> 2018-11-19 17:39:09,082 [WARNING] foremast.consts:validate_key_values:71 - [base] missing key "default_run_as_user", using None.
Traceback (most recent call last):
File "/tmp/workspace/pipes-scaling-policy/venv/lib/python3.6/site-packages/foremast/consts.py", line 158, in _convert_string_to_native
result = ast.literal_eval(value)
File "/usr/lib64/python3.6/ast.py", line 85, in literal_eval
return _convert(node_or_string)
File "/usr/lib64/python3.6/ast.py", line 84, in _convert
raise ValueError('malformed node or string: ' + repr(node))
ValueError: malformed node or string: {'bastion': [
> Unknown macro: {'start_port'}
], 'eureka': [\{'start_port': '8080', 'end_port': '8080', 'protocol': 'tcp'}], 'enabled_access_jenkinsqe_slave': [\{'start_port': '80', 'end_port': '80', 'protocol': 'tcp'}, \{'start_port': '443', 'end_port': '443', 'protocol': 'tcp'}, \{'start_port': '8080', 'end_port': '8080', 'protocol': 'tcp'}, \{'start_port': '8443', 'end_port': '8443', 'protocol': 'tcp'}], '$self': [\{'start_port': '0', 'end_port': '65535', 'protocol': 'tcp'}]}
> During handling of the above exception, another exception occurred:

The function in question is:

```
def _convert_string_to_native(value):
    """Convert a string to its native python type"""
    result = None

    try:
        result = ast.literal_eval(value)
    except (SyntaxError, ValueError):
        # Likely a string
        result = value.split(',')
    return result
```

`ast.literal_eval()` is expecting a string representation of a dict, but this parsing has already been done and the method is throwing ValueError when passed the actual dict.

**Comment by JessicaAlan** [comment]
Would adding 
```
if isinstance(value, str):
    return value
```
to the method fix the issue without impacting anything else?

**Comment by sijis** [comment]
That looks to be complaining around something within config.py, no?

What does the security group config line look like.

If I recall, when we parse the contents, via configparser, we always get strings (since we initially used an in style config). Your fix may work but we'd have to validate with ini and py config files.

**Comment by JessicaAlan** [comment]
```
        'default_securitygroup_rules': {
            "bastion": [
                {"start_port": "22", "end_port": "22", "protocol": "tcp"}
            ],
            "eureka": [
                {"start_port": "8080", "end_port": "8080", "protocol": "tcp"}
            ],
            "enabled_access_jenkinsqe_slave": [
                {"start_port": "80", "end_port": "80", "protocol": "tcp"},
                {"start_port": "443", "end_port": "443", "protocol": "tcp"},
                {"start_port": "8080", "end_port": "8080", "protocol": "tcp"},
                {"start_port": "8443", "end_port": "8443", "protocol": "tcp"}
            ],
            "$self": [
                {"start_port": "0", "end_port": "65535", "protocol": "tcp"}
            ],
        },
```

Per Steve "last I remember the config.py had some issue with spaces after commas in some of the “lists”, I forget if you needed them or couldn’t have them"

**Comment by jvasallo** [comment]
@JessicaAlan Noticing two issues with JSON lint in the above. Not sure if Copy Pasta.

L1: JSON requires double quotes instead of single quotes.
L16: You seem to have a trailing comma after the list as well.

Can you confirm locally? Basically you should be able to run:
`python -m json.tool <file>` 
and it shouldn't error out!

**Comment by JessicaAlan** [comment]
When I walk through the source code:

1. `find_config()` returns a dict, cast from a ConfigParser object
2. `validate_key_values()` extracts the `default_securitygroup_rules` dict from the config dict.
3. This dict is then passed to `_convert_string_to_native()`, resulting in the error.

https://github.com/foremast/foremast/blob/master/src/foremast/consts.py

**Comment by JessicaAlan** [comment]
@jvasallo I don't believe it's a JSON issue, that's from `config.py` and represents a literal dict.

**Comment by jvasallo** [comment]
@JessicaAlan I tried debugging this issue for bit now, but unfortunately I am not able to reproduce the issue with my own templates or the example you provided above. I spent a few attempts attempting to break this as well but still no luck; I did find another issue but it is not related to what you are reporting.

Are you by chance running the latest version locally and still able to reproduce? Something also to note, I am running Python 3.7.1 locally. I see you are running 3.6. Highly doubt the Python issue but can you also give that a shot?
Here is the command I used to test:
`ENV=stage REGION=us-east-1 GIT_REPO=service PROJECT=demo RUNWAY_DIR=/Users/user1/repos/demo/service/runway prepare-infrastructure`

Here is my config.py I used to test:
```
CONFIG = {
    'base': {
        'domain': 'example.com',
        'envs': 'stage',
        'regions': 'us-east-1',
        'gate_api_url': 'http://gate.spinnaker.example.com:8084',
        'templates_path': '/tmp/foremast-template-examples',
        'default_securitygroup_rules': {
            "servicedemoone": [
                {"start_port": "22", "end_port": "22", "protocol": "tcp"}
            ],
            "servicedemotwo": [
                {"start_port": "8080", "end_port": "8080", "protocol": "tcp"}
            ],
            "servicedemothree": [
                {"start_port": "80", "end_port": "80", "protocol": "tcp"},
                {"start_port": "443", "end_port": "443", "protocol": "tcp"},
                {"start_port": "8080", "end_port": "8080", "protocol": "tcp"},
                {"start_port": "8443", "end_port": "8443", "protocol": "tcp"}
            ],
            "$self": [
                {"start_port": "0", "end_port": "65535", "protocol": "tcp"}
            ],
        },
    },
    'formats': {
        'app': '{project}{repo}',
        'dns_elb': 'lb-{project}{repo}.{env}.{domain}',
    }
}
```

⚠️ I was able to find another issue, however. If you do a self reference *BEFORE* the SG is created, it could cause a:
`foremast.exceptions.SpinnakerTaskError: The following security groups do not exist: 'serviceappname' in 'env' vpc-id1234`

**Comment by jvasallo** [comment]
I did reproduce this issue and I believe it was resolved a while back with #328. I think you folks were running a later version of Python and Spinnaker at the time and that is why I couldn't reproduce.

Closing this for now, but let me know if this is not solved. :)
